### ISSUE TITLE
[Clarification request] TOTP-enabled accounts and smartphone client access

### ISSUE THREAD

**Issue by Wikinaut** [nextcloud/twofactor_totp#91]
* Nextcloud server 10.0.1
* Nextcloud Android client 1.3.1

After having enabled TOTP on my Nextcloud user account, I noticed that my Android shows strange behaviour, perhaps due to unsuccessful authentication/login to the account. For example, "SofortUpload" appeared to work, but when starting the App, my account password was cleared. (If you want, I can try to reproduce these steps.)

Please can you explain in detail:

* whether TOTP-enabled Nextcloud accounts can (how?) or cannot be used with clients like the Android client.

**Comment by ChristophWurst** [comment]
Hi,

yes, that is possible. See https://docs.nextcloud.com/server/11/user_manual/user_2fa.html and https://docs.nextcloud.com/server/11/user_manual/session_management.html#managing-devices.

> my account password was cleared.

What do you mean by that?

**Comment by Wikinaut** [comment]
First at all, probably I used an somehow older version of your app (about three weeks old) which did not show up the "backup codes", and díd not show the option for the device-specific passwords. 

* Have you added these features only recently ?

When using my Android, I noticed that 
* suddenly the connection was lost (after having enabled TOTP on the server, but without generating and adding device-password to my device). 
* Then I disabled TOTP on the server.
* Going back to my Android, I found that the server Url., and account user name were present, but
* password field was empty, so that
* I had to re-enter my password

I will not try to follow the intructions on your page and come back here.

Closing the issue now (reopening only, if the problem persists.)

**Comment by ChristophWurst** [comment]
> which did not show up the "backup codes"

That is a feature of Nextcloud 11. You won't see that on Nextcloud 10 ;-)

> and díd not show the option for the device-specific passwords

Those passwords are provided by the Nextcloud server, not this app. You should see them even if this app is not installed.

> Have you added these features only recently ?

-> No 😉 



> suddenly the connection was lost (after having enabled TOTP on the server, but without generating and adding device-password to my device).

How did you configure your Android client before? Did you use your login password or an app-specific password? You have to use the latter as soon as you enable 2FA. The login password is then not allowed for security reasons.

> password field was empty, so that
> I had to re-enter my password

This is strange. Maybe the android client resets the password if it detects authentication problems? Might be a bug, so worth filing an issue at the android repo. Could you please do that? Thanks

> I will not try to follow the intructions on your page and come back here.

I hope you meant *now* instead of *not* 😨

**Comment by Wikinaut** [comment]
I do not see the Backup code section!

**Comment by ChristophWurst** [comment]
> I do not see the Backup code section!

See comment above

**Comment by Wikinaut** [comment]
> This is strange. Maybe the android client resets the password if it detects authentication problems? Might be a bug, so worth filing an issue at the android repo. Could you please do that? Thanks

Looks so.

**Comment by Wikinaut** [comment]
> and díd not show the option for the device-specific passwords
> Those passwords are provided by the Nextcloud server, not this app. You should see them even if this app is not installed.

This was and is clear.

**Comment by Wikinaut** [comment]
Regarding your question,

> Did you use your login password or an app-specific password? 

Yes

> You have to use the latter as soon as you enable 2FA. The login password is then not allowed for security reasons.

This is clear, but was not so apparent. I think, your code could show an info text, when user enables TOTP "After enabling TOTP, all password-access enabled devices need to be supplied with a new device-specific password, see section ..."

**Comment by ChristophWurst** [comment]
> This is clear, but was not so apparent. I think, your code could show an info text, when user enables TOTP "After enabling TOTP, all password-access enabled devices need to be supplied with a new device-specific password, see section ..."

Agreed, that makes sense. I will try to improve this for NC12 because ideally that warning should be shown for all 2FA providers, so it has to go into the Nextcloud server and not specific apps. We documented that on the bottom of https://docs.nextcloud.com/server/11/user_manual/session_management.html#managing-devices but of course, people that don't read the documentation won't know about that.
Thanks for your feedback!

**Comment by Wikinaut** [comment]
I have three clients, i) Android ii) Linux iii) Mac . How to add device-passwords, do I have to generate for each device a new one ? (Then the text `the password is only shown once for security reasons` or the text above the device-password section should be improved `you can generate as many device-passwords as you need, but each will be shown only once.`)

**Comment by ChristophWurst** [comment]
I recommend to use one password per device/app.

Regarding the text, that's a valid point. Feel free to file an enhancement issue in the server repo and we can fix that easily :-)

**Comment by Wikinaut** [comment]
Another issue: indeed, my password was emptied in the Android Client, as soon as I activated TOTP (or thereafter).

On my Linux client (still owncloud Linux 2.2.4 because there is no other available for debian 8), this client is apparently still working! (No forced-logout!) 

How can this be ?

**Comment by ChristophWurst** [comment]
> On my Linux client (still owncloud Linux 2.2.4 because there is no other available for debian 8), this client is apparently still working! (No forced-logout!)

It might have an active PHP session. It should expire soon though.

**Comment by Wikinaut** [comment]
This would however not explain the effect on the Android, would you agree ?

**Comment by Wikinaut** [comment]
I stopped the client, and restartet, and now it pings me for a(the) password.

**Comment by ChristophWurst** [comment]
> This would however not explain the effect on the Android, would you agree ?

No. AFAIK the android client does not use sessions. But please file an issue in the android repo, I'm sure someone else knows more about the mobile client :-)

**Comment by Wikinaut** [comment]
@ChristophWurst regarding the emptied password, the current behaviour looks okay to me.
